Update README.md (#131)
cloudinary/cloudinary_android: README.md

Cloudinary Android SDK
======================

> [![Build Status](https://api.travis-ci.com/cloudinary/cloudinary-android.svg?branch=master)](https://app.travis-ci.com/github/cloudinary/cloudinary_android)

Reviewer: Yes, all SDKs have this. 
Everything that runs through Travis has it.
Just put the name of the Org/Repo in the URL and you'll get a status SVG, like this for Vue:

https://api.travis-ci.com/cloudinary/cloudinary-vue.svg?branch=master

> [![Build Status](https://api.travis-ci.com/cloudinary/cloudinary-android.svg?branch=master)](https://app.travis-ci.com/github/cloudinary/cloudinary_android)

Author: Fixed

## About
The Cloudinary Android SDK allows you to quickly and easily integrate your application with Cloudinary.

> Effortlessly optimize and transform your cloud's assets.

Reviewer: The template we created is designed for people who already know what Cloudinary is, and we provide the shortest explanation possible for people who accidentally got to this page directly.

### Additional documentation
This Readme provides basic installation and usage information.
For the complete documentation, see the [Android SDK Guide](https://cloudinary.com/documentation/android_integration).

## Table of Contents

> - [Key Features](#key-features)

Reviewer: This is something I, as a developer, would prefer seeing (A brief highlight of this SDK does) - that's what we agreed in the template

- [Compatibility](#Compatibility)
- [Installation](#installation)
- [Usage](#usage)
    - [Setup](#Setup)
    - [Transform and Optimize Assets](#Transform-and-Optimize-Assets)
    - [Uploading Asset](#Uploading)

## Key Features

> - [Image Transformation](https://cloudinary.com/documentation/android_image_manipulation) and [Video Transformation](https://cloudinary.com/documentation/android_video_manipulation) assets (links to docs).

Author: Done

## Compatibility
* Minimum Android SDK: Cloudinary requires a minimum API level of 19.

> * Compile Android SDK: Cloudinary requires you to compile against API 31 or later.

Reviewer 2: I have some wording suggestions, but I don't really understand what it means that we "require" level 19, but that you have to compile against 31?  What is it that you can do with level 19 after you've compiled against 31? Is 19 the level of Android that the end-users of the mobile app need to have?

Reviewer: My comment here is two fold:

1. We should use a table to make it clearer (as the template shows)
2. Required/Compile might confuse people, it might be very common for Android developers to use this terminology, but it still looks confusing.

Author: After talking to @patrick-tolosa we are removing the compile SDK version as it might confuse people 
After looking at the top SDK's for Android on git I'll only put the minimum SDK version required for our SDK
As its name implies the minimum SDK version means that lower version cannot compile/work with our SDK.

Author: This is another suggestion from another Android repo.
Cloudinary requires at minimum Java 8+ or Android API 19+.

Author: This is another suggestion from another repo
Cloudinary requires at minimum Java 8+ and Android API 19+.

Author: After talking with @patrick-tolosa we decided to use a table to be consistent with other SDK's

| Cloudinary SDK | Android SDK | 
|----------------|-------------|
|      1.x       |    > 14     |
|      2.x       |    > 19     |

## Installation

### Gradle Integration
Add the following dependency to your build.gradle:

`implementation 'com.cloudinary:cloudinary-android:1.30.0'`

> ### Manual Setup ######################################################################

Author: Done
Though I'm afraid people will not understand where the maven installation instructions end and the manual begins since I've swapped the titles into one "More Options"

Cloudinary Android SDK
======================
[![Build Status](https://api.travis-ci.com/cloudinary/cloudinary_android.svg?branch=master)](https://app.travis-ci.com/github/cloudinary/cloudinary_android)

## About
The Cloudinary Android SDK allows you to quickly and easily integrate your application with Cloudinary.
Effortlessly optimize and transform your cloud's assets.

### Additional documentation
This Readme provides basic installation and usage information.
For the complete documentation, see the [Android SDK Guide](https://cloudinary.com/documentation/android_integration).

## Table of Contents
- [Key Features](#key-features)
- [Compatibility](#Version Support)
- [Installation](#installation)
- [Usage](#usage)
    - [Setup](#Setup)

>     - [Transform and Optimize Assets](#Transforming and Optimizing Assets)

Reviewer: This is a broken link, please see how it looks like:
https://github.com/cloudinary/cloudinary_android/blob/update/readme/README.md

>     - [Uploading Asset](#Uploading Assets)

Reviewer: Also broken
https://github.com/cloudinary/cloudinary_android/blob/update/readme/README.md

        <version>1.30.0</version>
    </dependency>

## Usage

> ### Setup

Reviewer 3: Also, would replace the entire text, with steps. "Do 1,2,3" with less explanations. 
Would try to avoid alternate ways of doing the same thing. 
Remember - the main thing here is getting from 0-100 in minimal time. or in other words - having Cloudinary running in your code in no-time. 

Something like:
- Go to console (with link) 
- Copy `CLOUDINARY_URL`
- Paste in `AndroidManifest.xml`
- etc...

> ### Setup

Reviewer 3: Of course, at the end we should always put links to 'learn more' in the docs. 
But the important things is that we have a running code in the user's IDE.

> Each request for building a URL of a remote cloud resource must have the `cloud_name` parameter set.

Reviewer 2: I know this was already here, but I don't understand this sentence - what does it mean to build a URL of a "remote cloud resource"?

Author: I think remote cloud source = remote asset
an asset stored on the cloud and not locally.

Reviewer: I believe that was a hint to replace it with something easier to digest :) 
I'm sure @jackieros can offer an alternative phrase here.

> Setting the `cloud_name` parameter can be done either when initializing the library, or by using the CLOUDINARY_URL meta-data property in `AndroidManifest.xml`.

Reviewer 2: ```suggestion
You can set the `cloud_name` parameter either when initializing the library, or by using the CLOUDINARY_URL meta-data property in `AndroidManifest.xml`.
```

Reviewer 2: Just a general wording improvement. Wherever possible, instructions should be in 2nd person (you can) and not passive ("can be done...")

Author: Fixed, thank you

The entry point of the library is the `MediaManager` object. `MediaManager.init()` must be called before using the library, preferably in `Application.onCreate()`.
Here's an example of setting the configuration parameters programmatically in your `Applicaion.onCreate(`:

     Map config = new HashMap();
     config.put("cloud_name", "myCloudName");
     MediaManager.init(this, config);

> Alternatively, When using the meta-data property, no configuration is required:

Reviewer 2: ```suggestion
Alternatively, you can use the meta-data property. In that case, no configuration is required:
```

    MediaManager.init(this);

> The added property `AndroidManifest.xml`. Note: You should only include the `cloud_name` in the value, the api secret and key should be left out of the application.

Reviewer 2: Again, I know this was already there, but I don't think this readme was ever reviewed in the past. 
What does it mean to say "The added property `xxx`." as a sentence?  What are we trying to say here?

As for the note part, I suggest wording such as: 

"Only the `cloud_name` should be included.  Your API key and secret aren't necessary. 
**Note**: Your API secret should never be exposed in the application.

Author: The added property was referring to the meta-data you add the the file `AndroidManifest.xml` with the cloud name.
I'll change it to your suggestion as it looks better and more clear.

> ### Transform and Optimize Assets

Reviewer 2: ```suggestion
### Transforming and Optimizing Assets
```

Any image uploaded to Cloudinary can be transformed and embedded using powerful view helper methods:

The following example generates the url for accessing an uploaded `sample` image while transforming it to fill a 100x150 rectangle:

    MediaManager.get().url().transformation(new Transformation().width(100).height(150).crop("fill")).generate("sample.jpg")

Another example, embedding a smaller version of an uploaded image while generating a 90x90 face detection based thumbnail:

    MediaManager.get().url().transformation(new Transformation().width(90).height(90).crop("thumb").gravity("face")).generate("woman.jpg")

You can provide either a Facebook name or a numeric ID of a Facebook profile or a fan page.

Embedding a Facebook profile to match your graphic design is very simple:

    MediaManager.get().url().type("facebook").transformation(new Transformation().width(130).height(130).crop("fill").gravity("north_west")).generate("billclinton.jpg")

> Same goes for Twitter:

Reviewer 3: Would drop the twitter example

Author: Done

    MediaManager.get().url().type("twitter_name").generate("billclinton.jpg")

> ### Uploading

Reviewer 2: ```suggestion
### Uploading Assets
```

Reviewer 2: Changed the two headings to be consistent styles. Using gerunds (which are nouns) rather than verb headings, because all the H2s in the template are also nouns.

@patrick-tolosa - This should be followed at a template level.

The entry point for upload operations is the `MediaManager.get().upload()` call. All upload operations are dispatched to a background queue, with
a set of fully customizable rules and limits letting you choose when each upload request should actually run. Requests are automatically rescheduled to be
retried later if a recoverable error is encountered (e.g. network disconnections, timeouts).

> The upload results are dispatched asynchronously using `UploadCallback`. Global callbacks can be defined, as well as specific callbacks per request.

Reviewer 3: Way too much information.
Again, IMO, steps... 
After you get the user to make it work, you can explain the details.

    http://res.cloudinary.com/demo/image/upload/abcfrmo8zul1mafopawefg.jpg

You can also specify your own public ID:

    String requestId = MediaManager.get().upload(uri).unsigned("sample_preset").option("public_id", "sample_remote").dispatch();

Using `RequestUploadPolicy`, an upload request can be configured to run under specific circumstance, or within a chosen time window:

> The following examples uploads local Uri resource, configured to run immediately (the default), with a maximum of 7 retries, and only on an un-metered network (e.g. wifi):

Reviewer 2: In general, words with 'un' should not be hyphenated unless the word could have another meaning without the hyphen.  And from everything I see on Google, this term is always used without a hyphen, so we should revert this one.

Author: Done, thank you

For security reasons, mobile applications cannot contain the full account credentials, and so they cannot freely upload resources to the cloud.
Cloudinary provides two different mechanisms to enable end-users to upload resources without providing full credentials.

> ##### 1. Unsigned uploads using [Upload Presets.](http://cloudinary.com/blog/centralized_control_for_image_upload_image_size_format_thumbnail_generation_tagging_and_more)

Reviewer 2: This is linking to a 2014 blog post that probably has tons of no longer accurate info. In general, linking to blog posts is risky as the info in them is never updated.

Suggest linking here instead: android_image_and_video_upload#unsigned_upload

And if any info is missing there, feel free to open a DOC JIRA to improve that section.

Author: I've changed to the following link:
https://cloudinary.com/documentation/android_image_and_video_upload
